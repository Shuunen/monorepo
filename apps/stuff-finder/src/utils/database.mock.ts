import { functionReturningVoid, nbDaysInWeek, sleep } from '@shuunen/utils'
import type { Models } from 'appwrite'
import { vi } from 'vitest'
import type { ItemModel } from '../types/item.types'
import { mockItemModel } from './mock.utils'

export function mockFile(data: Partial<Models.File> = {}) {
  return {
    $createdAt: '2020-03-01T00:00:00.000Z',
    $id: 'some-image-file-uuid-xyz',
    $permissions: [],
    $updatedAt: '2021-08-01T00:00:00.000Z',
    bucketId: 'bucketA',
    chunksTotal: 1,
    chunksUploaded: 1,
    mimeType: 'image/jpeg',
    name: 'fileName-a.jpg',
    signature: 'signature-a',
    sizeOriginal: 123_456,
    ...data,
  } satisfies Models.File as Models.File
}

const createFile = vi.fn(async (bucketId: string, fileId: string, _file: File) => {
  await sleep(nbDaysInWeek)
  return { $id: fileId, bucketId, isThisMockedDataFromMock: true }
})

const deleteFile = vi.fn(async (bucketId: string, fileId: string, _file: File) => {
  await sleep(nbDaysInWeek)
  return { $id: fileId, bucketId, isThisMockedDataFromMock: true }
})

const listFiles = vi.fn(async (_bucketId: string, _queries: [{ limit: number }, { offset: number }]) => {
  await sleep(nbDaysInWeek)
  return { files: [], total: 0 } satisfies Models.FileList as Models.FileList
})

// oxlint-disable-next-line max-params
const createDocument = vi.fn(async (databaseId: string, collectionId: string, documentId: string, data: object) => {
  await sleep(nbDaysInWeek)
  const item = mockItemModel({ $collectionId: collectionId, $databaseId: databaseId, $id: documentId, ...data })
  return item satisfies Models.Document as Models.Document
})

const deleteDocument = vi.fn(async (databaseId: string, collectionId: string, documentId: string) => {
  await sleep(nbDaysInWeek)
  return { $id: documentId, collectionId, databaseId, isThisMockedDataFromMock: true }
})

const listDocuments = vi.fn(async (_databaseId: string, _collectionId: string) => {
  await sleep(nbDaysInWeek)
  return { documents: [], total: 0 } satisfies Models.DocumentList<ItemModel> as Models.DocumentList<ItemModel>
})

// oxlint-disable-next-line max-params
const updateDocument = vi.fn(async (databaseId: string, collectionId: string, documentId: string, data: object) => {
  await sleep(nbDaysInWeek)
  const item = mockItemModel({ $collectionId: collectionId, $databaseId: databaseId, $id: documentId, ...data })
  return item satisfies Models.Document as Models.Document
})

class Databases {
  createDocument = createDocument
  deleteDocument = deleteDocument
  listDocuments = listDocuments
  updateDocument = updateDocument
  constructor(client?: Client) {
    if (client) functionReturningVoid()
  }
}

// oxlint-disable-next-line max-classes-per-file
class Client {
  constructor() {
    functionReturningVoid()
  }
  setProject(project: string) {
    if (project) functionReturningVoid()
    return this
  }
}

class Storage {
  createFile = createFile
  deleteFile = deleteFile
  listFiles = listFiles
  constructor(client?: Client) {
    if (client) functionReturningVoid()
  }
}

const Query = {
  limit: vi.fn((nb: number) => ({ isThisMockedDataFromMock: true, limit: nb })),
  offset: vi.fn((nb: number) => ({ isThisMockedDataFromMock: true, offset: nb })),
}

function reset() {
  createDocument.mockClear()
  createFile.mockClear()
  listFiles.mockClear()
  deleteDocument.mockClear()
  deleteFile.mockClear()
  listDocuments.mockClear()
  updateDocument.mockClear()
  Query.limit.mockClear()
  Query.offset.mockClear()
}

// oxlint-disable-next-line sort-keys
export const databaseMock = {
  // biome-ignore lint/style/useNamingConvention: I cant change this, it's part of the Appwrite SDK
  appwrite: { Client, Databases, Query, Storage },
  createDocument,
  createFile,
  deleteDocument,
  deleteFile,
  listDocuments,
  listFiles,
  // biome-ignore lint/style/useNamingConvention: I cant change this, it's part of the Appwrite SDK
  Query,
  reset,
  updateDocument,
}
