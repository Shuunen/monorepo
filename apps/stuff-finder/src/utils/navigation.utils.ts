import type { NavigateFunction } from 'react-router-dom'

let navigateFunction: NavigateFunction | undefined = undefined

/**
 * Sets the global navigation function to be used for programmatic navigation
 * @param aFunction - The navigation function to set, typically provided by a routing library but can be useful for testing or custom navigation logic.
 */
export function setNavigate(aFunction: NavigateFunction) {
  navigateFunction = aFunction
}

/**
 * Navigates to the specified path
 * @param path - The target path to navigate to.
 * @param replace - Optional. If true, replaces the current entry in the history stack instead of adding a new one. Defaults to false.
 */
export function navigate(path: string, replace = false) {
  if (navigateFunction) navigateFunction(path, { replace })
}
