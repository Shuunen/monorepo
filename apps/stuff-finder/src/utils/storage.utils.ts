import { storage } from '@shuunen/shuutils'

storage.prefix = '@shuunen/stuff-finder_'

export { storage }
