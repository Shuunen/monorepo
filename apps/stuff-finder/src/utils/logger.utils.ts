import { Logger } from '@shuunen/utils'

export const logger = new Logger()
