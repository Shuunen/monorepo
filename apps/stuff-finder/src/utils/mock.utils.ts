import { nbDaysInWeek, sleep } from '@shuunen/utils'
import { vi } from 'vitest'
import type { Item, ItemModel } from '../types/item.types'
import { defaultSound } from '../types/sounds.types'
import { defaultStatus } from '../types/status.types'
import { defaultTheme } from '../types/theme.types'
import type { State } from './state.utils'

export function mockItem(data: Partial<Item> = {}) {
  return {
    $createdAt: '2025-07-16T13:42:26.000Z',
    $id: 'rec234',
    barcode: 'barcode B',
    box: 'B (usb & audio)',
    brand: 'brand B',
    details: 'details B',
    drawer: 2,
    isPrinted: false,
    name: 'name B',
    photos: ['some-uuid', 'https://some.url/to/image.jpg'],
    price: 42,
    reference: 'reference B',
    status: 'bought',
    ...data,
  } satisfies Item as Item
}

export function mockItemModel(data: Partial<ItemModel> = {}) {
  return {
    ...mockItem(),
    $collectionId: 'col234',
    $createdAt: '2020-03-01T00:00:00.000Z',
    $databaseId: 'db234',
    $permissions: [],
    $sequence: 1,
    $updatedAt: '2021-08-01T00:00:00.000Z',
    box: 'B (usb & audio)',
    drawer: 2,
    ...data,
  } satisfies ItemModel as ItemModel
}

export function mockState(data: Partial<State> = {}) {
  return {
    credentials: { bucketId: 'bucketA', collectionId: 'collectionA', databaseId: 'databaseA', wrap: 'wrapA' },
    display: 'list',
    items: [] satisfies Item[],
    itemsTimestamp: Date.now(),
    sound: defaultSound,
    status: defaultStatus,
    theme: defaultTheme,
    ...data,
  } satisfies State
}

export const mockFetch = vi.fn(async (input: RequestInfo | URL, options?: RequestInit) => {
  await sleep(nbDaysInWeek)
  return {
    blob: async () => {
      await sleep(nbDaysInWeek)
      return { input, options }
    },
  } as unknown as Promise<Response>
})
