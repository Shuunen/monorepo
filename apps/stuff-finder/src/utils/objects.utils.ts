import { arrayUnique, clone } from '@monorepo/utils'

/**
 * Sort lists
 * @param record the lists to sort
 * @returns the sorted list... unexpected right ?
 */
export function sortListsEntries<Type = Record<string, string[]>>(record: Type) {
  const output = clone<Type>(record)
  // @ts-expect-error problem with Object.entries
  const entries = Object.entries<string[]>(output)
  /* v8 ignore next -- @preserve */
  for (const [name, values] of entries)
    if (Array.isArray(values)) {
      // oxlint-disable-next-line new-cap
      const updatedValues = ['', ...values.toSorted((valueA: string, valueB: string) => Intl.Collator().compare(valueA, valueB))]
      ;(output as Record<string, string[]>)[name] = arrayUnique(updatedValues)
    }
  return output
}
