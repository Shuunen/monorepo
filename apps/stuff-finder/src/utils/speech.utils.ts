import { Result } from '@shuunen/utils'
import { logger } from './logger.utils'
import { state } from './state.utils'

/**
 * Listen user speech
 * @param onSuccess the function to call when the speech is recognized
 */
// oxlint-disable-next-line max-lines-per-function
export function listenUserSpeech(onSuccess: (transcript: string, confidence: number) => void) {
  let isSuccess = false
  const recognition = new (globalThis.webkitSpeechRecognition || globalThis.SpeechRecognition)()
  recognition.lang = 'fr-FR'
  recognition.interimResults = false
  recognition.maxAlternatives = 1
  /**
   * Called when the speech is recognized
   * @param event the speech recognition event
   * @returns the result of the recognition
   */
  recognition.onresult = (event: SpeechRecognitionEvent) => {
    isSuccess = true
    const [result] = Array.from(event.results.item(event.resultIndex))
    if (!result) return Result.error('no recognition result found')
    onSuccess(result.transcript, result.confidence)
    return Result.ok('speech recognized successfully')
  }
  /**
   * Called when the speech is not recognized
   */
  recognition.onnomatch = () => {
    logger.warn('speech : no match found')
    state.sound = 'error'
  }
  /**
   * Called when the speech recognition end
   */
  recognition.onend = () => {
    state.sound = isSuccess ? 'stop' : 'error'
  }
  // start the speech recognition, listen to the user
  state.sound = 'start'
  recognition.start()
}
