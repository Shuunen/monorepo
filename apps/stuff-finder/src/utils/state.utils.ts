import { createState, debounce, isBrowserEnvironment, isTestEnvironment } from '@shuunen/shuutils'
import { defaultCredentials } from '../constants'
import type { Item } from '../types/item.types'
import { defaultSound } from '../types/sounds.types'
import { type AppStatus, defaultStatus } from '../types/status.types'
import { type Display, defaultTheme } from '../types/theme.types'
import { navigate } from './navigation.utils'
import { storage } from './storage.utils'

/* c8 ignore start */

/**
 * Handle the status change
 * @param status the new status
 */
function onStatusChangeSync(status: AppStatus) {
  if (isTestEnvironment()) return
  if (status === 'settings-required') navigate('/settings')
  if (status === 'ready' && document.location.pathname.includes('/settings')) navigate('/')
}

const laptopWidth = 1500

const defaultDisplay: Display = isBrowserEnvironment() && globalThis.screen.width < laptopWidth ? 'list' : 'card'

/* c8 ignore stop */

export const { state, watchState } = createState(
  {
    credentials: defaultCredentials,
    display: defaultDisplay,
    items: [] as Item[],
    /** timestamp of the last time items were fetched, in milliseconds */
    itemsTimestamp: 0,
    sound: defaultSound,
    status: defaultStatus,
    /** the display theme of the item list : card or list */
    theme: defaultTheme,
  },
  storage,
  ['credentials', 'display', 'items', 'itemsTimestamp', 'theme'], // avoid status persistence
)

const statusDelay = 300

const onStatusChange = debounce(onStatusChangeSync, statusDelay)

watchState('status', () => {
  void onStatusChange(state.status)
})

export type State = typeof state
