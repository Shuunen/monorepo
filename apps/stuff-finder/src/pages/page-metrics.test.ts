import { describe, expect, it } from 'vitest'
import type { Item } from '../types/item.types'
import { mockItem } from '../utils/mock.utils'
import { calculateBasicMetrics, calculateBoxAnalysis, calculateMetrics, calculateStatusCounts, formatCurrency, formatPercentage, getItemsNotPrinted, getItemsWithoutLocation, getItemsWithoutPhoto, getItemsWithoutPrice, getToGiveItems, getTopValueItems } from './page-metrics.utils'

const mockItems: Item[] = [
  mockItem({
    $id: '1',
    box: 'A (apple)',
    brand: 'Apple',
    details: 'iPhone 12',
    isPrinted: true,
    name: 'iPhone 12',
    price: 500,
    reference: 'IP12',
    status: 'bought',
  }),
  mockItem({
    $id: '2',
    box: 'B (usb & audio)',
    brand: 'Sony',
    details: 'Wireless headphones',
    isPrinted: false,
    name: 'WH-1000XM4',
    price: 300,
    reference: 'WH1000XM4',
    status: 'for-sell',
  }),
  mockItem({
    $id: '3',
    box: 'A (apple)',
    brand: 'Apple',
    details: 'iPad Pro',
    isPrinted: true,
    name: 'iPad Pro',
    price: 0, // No price
    reference: 'IPADPRO',
    status: 'bought',
  }),
  mockItem({
    $id: '4',
    box: '',
    brand: '',
    details: 'Old cable',
    isPrinted: false,
    name: 'USB Cable',
    photos: [], // No photos
    price: -1, // Invalid price
    reference: 'USBCABLE',
    status: 'lost',
  }),
  mockItem({
    $id: '5',
    box: 'C (couteau)',
    brand: 'Generic',
    details: 'Power cable',
    isPrinted: true,
    name: 'Power Cable',
    photos: undefined, // No photos
    price: 10,
    reference: 'POWERCABLE',
    status: 'bought',
  }),
]

describe('page-metrics.utils', () => {
  it('calculateBasicMetrics A should calculate correct basic metrics', () => {
    const result = calculateBasicMetrics(mockItems)

    expect(result).toMatchInlineSnapshot(`
      {
        "totalItems": 5,
        "totalValue": 810,
      }
    `)
  })

  it('calculateStatusCounts A should count items by status correctly', () => {
    const result = calculateStatusCounts(mockItems)

    expect(result).toMatchInlineSnapshot(`
      {
        "bought": 3,
        "for-sell": 1,
        "lost": 1,
      }
    `)
  })

  it('calculateBoxAnalysis A should analyze box distribution correctly', () => {
    const result = calculateBoxAnalysis(mockItems)
    expect(result).toMatchSnapshot()
  })

  it('getTopValueItems A should return top valued items', () => {
    const result = getTopValueItems(mockItems)
    expect(result).toMatchSnapshot()
  })

  it('calculateMetrics A should return complete metrics object', () => {
    const result = calculateMetrics(mockItems)
    expect(result.totalItems).toMatchInlineSnapshot(`5`)
    expect(result.totalValue).toMatchInlineSnapshot(`810`)
    expect(result.topValueItems, 'topValueItems length').toHaveLength(3)
    expect(Object.keys(result.boxAnalysis), 'boxAnalysis length').toHaveLength(4)
    expect(Object.keys(result.statusCounts), 'statusCounts length').toHaveLength(3)
  })

  it('formatCurrency A should format currency correctly', () => {
    expect(formatCurrency(100)).toMatchInlineSnapshot(`"100 €"`)
    expect(formatCurrency(99.99)).toMatchInlineSnapshot(`"100 €"`)
    expect(formatCurrency(0)).toMatchInlineSnapshot(`"0 €"`)
  })

  it('formatPercentage A should format percentage correctly', () => {
    expect(formatPercentage(50, 100)).toMatchInlineSnapshot(`"50.0 %"`)
    expect(formatPercentage(1, 3)).toMatchInlineSnapshot(`"33.3 %"`)
    expect(formatPercentage(0, 0)).toMatchInlineSnapshot(`"0.0 %"`)
    expect(formatPercentage(0, 100)).toMatchInlineSnapshot(`"0.0 %"`)
  })

  it('calculateBasicMetrics B should handle empty array', () => {
    const result = calculateBasicMetrics([])
    expect(result).toMatchInlineSnapshot(`
      {
        "totalItems": 0,
        "totalValue": 0,
      }
    `)
  })

  it('calculateStatusCounts B should handle empty array', () => {
    const result = calculateStatusCounts([])

    expect(result).toMatchInlineSnapshot(`{}`)
  })

  it('getTopValueItems B should handle items without prices', () => {
    const itemsWithoutPrices: Item[] = [
      mockItem({
        $id: '1',
        name: 'Item 1',
        price: 0,
        reference: 'REF1',
        status: 'bought',
      }),
    ]
    const result = getTopValueItems(itemsWithoutPrices)
    expect(result).toMatchInlineSnapshot(`[]`)
  })

  it('getToGiveItemsCount A should return 0 when no items have to-give status', () => {
    const result = getToGiveItems(mockItems)
    expect(result).toMatchInlineSnapshot(`[]`)
  })

  it('getToGiveItemsCount B should count items with to-give status', () => {
    const itemsWithToGive = [
      mockItem({
        $id: '1',
        name: 'Item to give 1',
        reference: 'GIVE1',
        status: 'to-give',
      }),
      mockItem({
        $id: '2',
        name: 'Item to give 2',
        reference: 'GIVE2',
        status: 'to-give',
      }),
      mockItem({
        $id: '3',
        name: 'Item not to give',
        reference: 'NOGIVE',
        status: 'bought',
      }),
    ]
    const result = getToGiveItems(itemsWithToGive)
    expect(result).toMatchSnapshot()
  })

  it('getItemsNotPrinted A should return items that are not printed', () => {
    const result = getItemsNotPrinted(mockItems)

    expect(result.map(item => item.$id)).toMatchInlineSnapshot(`
      [
        "2",
        "4",
      ]
    `)
  })

  it('getItemsWithoutLocation A should return items without box location', () => {
    const result = getItemsWithoutLocation(mockItems)

    expect(result.map(item => item.$id)).toMatchInlineSnapshot(`
      [
        "4",
      ]
    `)
  })

  it('getItemsWithoutPrice A should return items without valid price', () => {
    const result = getItemsWithoutPrice(mockItems)

    expect(result.map(item => item.$id)).toMatchInlineSnapshot(`
      [
        "3",
        "4",
      ]
    `)
  })

  it('getItemsWithoutPhoto A should return items without photos', () => {
    const result = getItemsWithoutPhoto(mockItems)

    expect(result.map(item => item.$id)).toMatchInlineSnapshot(`
      [
        "4",
        "5",
      ]
    `)
  })

  it('calculateMetrics B should include new list properties', () => {
    const result = calculateMetrics(mockItems)

    expect(result.itemsNotPrinted).toHaveLength(2)
    expect(result.itemsWithoutLocation).toHaveLength(1)
    expect(result.itemsWithoutPhoto).toHaveLength(2)
    expect(result.itemsWithoutPrice).toHaveLength(2)
    expect(result.itemsNotPrinted[0].$id).toMatchInlineSnapshot(`"2"`)
    expect(result.itemsWithoutLocation[0].$id).toMatchInlineSnapshot(`"4"`)
    expect(result.itemsWithoutPhoto[0].$id).toMatchInlineSnapshot(`"4"`)
    expect(result.itemsWithoutPrice[0].$id).toMatchInlineSnapshot(`"3"`)
  })
})
