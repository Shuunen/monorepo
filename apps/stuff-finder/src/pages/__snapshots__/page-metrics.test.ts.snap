// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page-metrics.utils > calculateBoxAnalysis A should analyze box distribution correctly 1`] = `
{
  "A (apple)": {
    "count": 2,
    "totalValue": 500,
  },
  "B (usb & audio)": {
    "count": 1,
    "totalValue": 300,
  },
  "C (couteau)": {
    "count": 1,
    "totalValue": 10,
  },
  "Unknown": {
    "count": 1,
    "totalValue": 0,
  },
}
`;

exports[`page-metrics.utils > getToGiveItemsCount B should count items with to-give status 1`] = `
[
  {
    "$createdAt": "2025-07-16T13:42:26.000Z",
    "$id": "1",
    "barcode": "barcode B",
    "box": "B (usb & audio)",
    "brand": "brand B",
    "details": "details B",
    "drawer": 2,
    "isPrinted": false,
    "name": "Item to give 1",
    "photos": [
      "some-uuid",
      "https://some.url/to/image.jpg",
    ],
    "price": 42,
    "reference": "GIVE1",
    "status": "to-give",
  },
  {
    "$createdAt": "2025-07-16T13:42:26.000Z",
    "$id": "2",
    "barcode": "barcode B",
    "box": "B (usb & audio)",
    "brand": "brand B",
    "details": "details B",
    "drawer": 2,
    "isPrinted": false,
    "name": "Item to give 2",
    "photos": [
      "some-uuid",
      "https://some.url/to/image.jpg",
    ],
    "price": 42,
    "reference": "GIVE2",
    "status": "to-give",
  },
]
`;

exports[`page-metrics.utils > getTopValueItems A should return top valued items 1`] = `
[
  {
    "$createdAt": "2025-07-16T13:42:26.000Z",
    "$id": "1",
    "barcode": "barcode B",
    "box": "A (apple)",
    "brand": "Apple",
    "details": "iPhone 12",
    "drawer": 2,
    "isPrinted": true,
    "name": "iPhone 12",
    "photos": [
      "some-uuid",
      "https://some.url/to/image.jpg",
    ],
    "price": 500,
    "reference": "IP12",
    "status": "bought",
  },
  {
    "$createdAt": "2025-07-16T13:42:26.000Z",
    "$id": "2",
    "barcode": "barcode B",
    "box": "B (usb & audio)",
    "brand": "Sony",
    "details": "Wireless headphones",
    "drawer": 2,
    "isPrinted": false,
    "name": "WH-1000XM4",
    "photos": [
      "some-uuid",
      "https://some.url/to/image.jpg",
    ],
    "price": 300,
    "reference": "WH1000XM4",
    "status": "for-sell",
  },
  {
    "$createdAt": "2025-07-16T13:42:26.000Z",
    "$id": "5",
    "barcode": "barcode B",
    "box": "C (couteau)",
    "brand": "Generic",
    "details": "Power cable",
    "drawer": 2,
    "isPrinted": true,
    "name": "Power Cable",
    "photos": undefined,
    "price": 10,
    "reference": "POWERCABLE",
    "status": "bought",
  },
]
`;
